### ISSUE TITLE
Support for OpenJDK 15

### ISSUE THREAD

**Issue by petebankhead** [qupath/qupath#615]
## Bug report

**Describe the bug**
OpenSlide doesn't work on Windows when QuPath is build with OpenJDK 15.
This appears to come from a change in jpackage - see https://forum.image.sc/t/mrxs-format-not-read-after-building-qupath-0-2-3-with-tensorflow/44122/6

**Expected behavior**
OpenSlide works as before.

**Desktop (please complete the following information):**
 - OS: Windows
 - QuPath Version: Future builds

**Comment by juliangilbey** [comment]
Here's another OpenJDK 15 issue; after building with OpenJDK 15, I get the following error message (on a Linux system):

```
euler:~/qupath (master) $ build/dist/QuPath-0.2.3/bin/QuPath-0.2.3 
Error opening "/home/jdg/qupath/build/dist/QuPath-0.2.3/lib/app/QuPath-0.2.cfg" file: No such file or directory
```

This does not occur when the package is built with OpenJDK 14, so somewhere the path is getting messed up. 

Best wishes,

   Julian

**Comment by petebankhead** [comment]
Ah thanks, good to know. Sounds like it is related to things now being rearranged, but I had been optimistically hoping it would be Windows-only.

**Comment by petebankhead** [comment]
Just for the full set, it fails on macOS too... albeit for a different reason (version numbers that start with 0 seem not to be supported).

**Comment by anotherBunchOfAtoms** [comment]
@juliangilbey I managed to solve this issue by duplicating QuPath-0.2.3.cfg and renaming it to QuPath-0.2.cfg
Best wishes,
Paul

**Comment by anotherBunchOfAtoms** [comment]
however I do get the following error when starting qupath:

`[JavaFX Application Thread] [ERROR] q.l.i.s.o.OpenslideServerBuilder - Could not load OpenSlide native libraries
java.lang.UnsatisfiedLinkError: no openslide-jni in java.library.path: /usr/java/packages/lib:/usr/lib64:/lib64:/lib:/usr/lib
        at java.base/java.lang.ClassLoader.loadLibrary(Unknown Source)
        at java.base/java.lang.Runtime.loadLibrary0(Unknown Source)
        at java.base/java.lang.System.loadLibrary(Unknown Source)
        at org.openslide.OpenSlideJNI.<clinit>(OpenSlideJNI.java:55)
        at org.openslide.OpenSlide.<clinit>(OpenSlide.java:53)
        at qupath.lib.images.servers.openslide.OpenslideServerBuilder.<clinit>(OpenslideServerBuilder.java:88)
        at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
        at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(Unknown Source)
        at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(Unknown Source)
        at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Unknown Source)
        at java.base/java.lang.reflect.Constructor.newInstance(Unknown Source)
        at java.base/java.util.ServiceLoader$ProviderImpl.newInstance(Unknown Source)
        at java.base/java.util.ServiceLoader$ProviderImpl.get(Unknown Source)
        at java.base/java.util.ServiceLoader$3.next(Unknown Source)
        at qupath.lib.images.servers.ImageServerProvider.getInstalledImageServerBuilders(ImageServerProvider.java:104)
        at qupath.lib.gui.QuPathGUI.refreshExtensions(QuPathGUI.java:1601)
        at qupath.lib.gui.QuPathGUI.<init>(QuPathGUI.java:1059)
        at qupath.lib.gui.QuPathApp.start(QuPathApp.java:60)
        at com.sun.javafx.application.LauncherImpl.lambda$launchApplication1$9(LauncherImpl.java:846)
        at com.sun.javafx.application.PlatformImpl.lambda$runAndWait$12(PlatformImpl.java:455)
        at com.sun.javafx.application.PlatformImpl.lambda$runLater$10(PlatformImpl.java:428)
        at java.base/java.security.AccessController.doPrivileged(Unknown Source)
        at com.sun.javafx.application.PlatformImpl.lambda$runLater$11(PlatformImpl.java:427)
        at com.sun.glass.ui.InvokeLaterDispatcher$Future.run(InvokeLaterDispatcher.java:96)
        at com.sun.glass.ui.gtk.GtkApplication._runLoop(Native Method)
        at com.sun.glass.ui.gtk.GtkApplication.lambda$runLoop$11(GtkApplication.java:277)
        at java.base/java.lang.Thread.run(Unknown Source)
17:02:49.461 [JavaFX Application Thread] [INFO ] q.l.i.s.o.OpenslideServerBuilder - If you want to use OpenSlide, you'll need to get the native libraries (either building from source or with a packager manager)
and add them to your system PATH, including openslide-jni.
17:02:49.798 [qupathgui-1] [INFO ] qupath.lib.gui.QuPathGUI - Performing update check...
17:02:49.798 [JavaFX Application Thread] [INFO ] qupath.lib.gui.QuPathApp - Starting QuPath with parameters: []
17:02:52.272 [qupathgui-1] [INFO ] qupath.lib.gui.QuPathGUI - Current version 0.2.3, latest stable release 0.2.3 - nothing to update`

**Comment by petebankhead** [comment]
@anotherBunchOfAtoms if you built it with OpenJDK 15, it's likely because `jpackage` puts files in a different place than it did with OpenJDK 14... and so QuPath can't find them. We'll need to find a way to address that, but haven't yet.

In the meantime, you need to build QuPath with OpenJDK 14 if you want OpenSlide to work.

**Comment by anotherBunchOfAtoms** [comment]
@petebankhead thank you for your response! I see what you mean. I think for now Bio-Formats is enough for me. I had some errors building with adoptopenjdk-14-hotspot so I built it with 15 (plus this little workaround for the cfg file to get it to run)

**Comment by petebankhead** [comment]
Ah ok, you might find shifting any native libraries (in the same way as the cfg file) can also do the job.

The underlying problem is that the Java library path needs to be set to find `openslide-jni`. Usually this 'just works' because the necessary library is found within the 'current' directory at the time QuPath is launched (through whatever magic jpackage has applied to set the current directory). You can also specify the library path in the `.cfg` file... but that may be more complicated.

You might need to install OpenSlide through a package manager too, thanks to a separate Linux-related problem regarding how sub-dependencies are found.... https://github.com/qupath/qupath/issues/629

Not sure how much sense that makes, and I haven't looked into what has changed in Linux - but the summary is that moving things around may well be the solution to get things going with Java 15.

**Comment by anotherBunchOfAtoms** [comment]
thanks for the explanation! I'll take a look at issue #629 too

**Comment by ertugrulcetin** [comment]
@anotherBunchOfAtoms could you please provide more about how you managed to build with JDK 15. I'm getting `UnsatisfiedLinkError` for my app (which contains native libs), and `jpackage` puts natives somewhere and does not add my native libs as well.

**Comment by petebankhead** [comment]
Quick update: there are no longer any plans to make QuPath OpenJDK 15-compatible. The next release is intended to be >= March 2021, by which time OpenJDK 16 should be available and there is a risk things will have changed again. However, jpackage is an intended feature for 16 so should be more stable... and worth updating for.

In the meantime, QuPath v0.2.* is only intended to be built with OpenJDK 14. QuPath's code should be compatible with Java 11; OpenJDK 14 is required only for jpackage. From an IDE, it should run with other Java versions.

**Comment by petebankhead** [comment]
Closing this, as the build script fixes are now included for v0.3.
(Note that we're skipping OpenJDK 15 and going straight to 16)
